flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/UserFunctionsTypeHelper.java

```diff
@@ -0,0 +1,176 @@
+
+		TypeInformation<ACC> userProvidedType = aggregateFunction.getAccumulatorType();
+		if (userProvidedType != null) {
+			return userProvidedType;
+		} else if (scalaType != null) {
+			return scalaType;
+		} else {
+			return TypeExtractor.createTypeInfo(
+				aggregateFunction,
+				UserDefinedAggregateFunction.class,
+				aggregateFunction.getClass(),
+				1);
+		}
+	}
+
+	/**
+	 * Tries to infer the TypeInformation of an AggregateFunction's accumulator type.
+	 *
+	 * @param tableFunction The TableFunction for which the accumulator type is inferred.
+	 * @return The inferred accumulator type of the AggregateFunction.
+	 */
+	public static <T> TypeInformation<T> getReturnTypeOfTableFunction(TableFunction<T> tableFunction) {
+		return getReturnTypeOfTableFunction(tableFunction, null);
+	}
+
+	/**
+	 * Tries to infer the TypeInformation of an AggregateFunction's accumulator type.
+	 *
+	 * @param tableFunction The TableFunction for which the accumulator type is inferred.
+	 * @param scalaType The implicitly inferred type of the accumulator type.
+	 * @return The inferred accumulator type of the AggregateFunction.
+	 */
+	public static <T> TypeInformation<T> getReturnTypeOfTableFunction(
+			TableFunction<T> tableFunction,
+			TypeInformation<T> scalaType) {
+
+		TypeInformation<T> userProvidedType = tableFunction.getResultType();
+		if (userProvidedType != null) {
+			return userProvidedType;
+		} else if (scalaType != null) {
+			return scalaType;
+		} else {
+			return TypeExtractor.createTypeInfo(
+				tableFunction,
+				TableFunction.class,
+				tableFunction.getClass(),
+				0);
+		}
+	}
+
+	/**
+	 * Checks if a user-defined function can be easily instantiated.
+	 */
+	public static void validateInstantiation(Class<?> clazz) {
+		if (!InstantiationUtil.isPublic(clazz)) {
+			throw new ValidationException(String.format("Function class %s is not public.", clazz.getCanonicalName()));
+		} else if (!InstantiationUtil.isProperClass(clazz)) {
+			throw new ValidationException(String.format(
+				"Function class %s is no proper class," +
+					" it is either abstract, an interface, or a primitive type.", clazz.getCanonicalName()));
+		} else if (InstantiationUtil.isNonStaticInnerClass(clazz)) {
+			throw new ValidationException(String.format(
+				"The class %s is an inner class, but not statically accessible.", clazz.getCanonicalName()));
+		}
+	}
+
+	/**
+	 * Check whether this is a Scala object. It is forbidden to use {@link TableFunction} implemented
+	 * by a Scala object, since concurrent risks.
+	 */
+	public static void validateNotSingleton(Class<?> clazz) {
+		// TODO it is not a good way to check singleton. Maybe improve it further.
+		if (Arrays.stream(clazz.getFields()).anyMatch(f -> f.getName().equals("MODULE$"))) {
+			throw new ValidationException(String.format(
+				"TableFunction implemented by class %s " +
+					"is a Scala object, it is forbidden since concurrent risks.", clazz.getCanonicalName()));
```

Comment:
I know it's not in the scope of the PR but this should be `this is forbidden because of concurrency problems when using them`. Or something like it.